src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,258 @@
+        DetailAST curNode = root;
+        while (curNode != null && curNode.getLineNo() <= lineNumber) {
+            if (curNode.getLineNo() == lineNumber && curNode.getColumnNo() == columnNumber
+                    && curNode.getType() != TokenTypes.IDENT) {
+                result.add(curNode);
+            }
+            DetailAST toVisit = curNode.getFirstChild();
+            while (curNode != null && toVisit == null) {
+                toVisit = curNode.getNextSibling();
+                if (toVisit == null) {
+                    curNode = curNode.getParent();
+                }
+            }
+
+            curNode = toVisit;
+        }
+        return result;
+    }
+
+    /**
+     * Returns relative xpath query for given ast element from root.
+     * @param root {@code DetailAST} root element
+     * @param ast {@code DetailAST} ast element
+     * @return relative xpath query for given ast element from root
+     */
+    private static String getXpathQuery(DetailAST root, DetailAST ast) {
+        String result = "";
+        DetailAST cur = ast;
+        while (cur != root) {
+            String curNodeQuery = '/' + TokenUtils.getTokenName(cur.getType());
+            final DetailAST identAst = cur.findFirstToken(TokenTypes.IDENT);
+            if (identAst != null) {
+                curNodeQuery += "[@text='" + identAst.getText() + "']";
+            }
+            result = curNodeQuery + result;
+            cur = cur.getParent();
+        }
+        return result;
+    }
+
+    /**
+     * Checks if the given ast element has unique {@code TokenTypes} among siblings.
+     * @param firstSibling {@code DetailAST} first sibling
+     * @param ast {@code DetailAST} ast element
+     * @return if the given ast element has unique {@code TokenTypes} among siblings
+     */
+    private static boolean hasNoSiblingWithSameTokenType(DetailAST firstSibling, DetailAST ast) {
+        DetailAST cur = firstSibling;
+        boolean result = true;
+        while (cur != null) {
+            if (cur != ast && cur.getType() == ast.getType()) {
+                result = false;
+                break;
+            }
+            cur = cur.getNextSibling();
+        }
+        return result;
+    }
+
+    /**
+     * To be sure that generated xpath query will return exactly required ast element, the element
+     * should be checked for uniqueness. If ast element has {@link TokenTypes#IDENT} as the child
+     * or there is no sibling with the same {@code TokenTypes} then element is supposed to be
+     * unique. This method finds if {@code DetailAst} element is unique.
+     * @param ast {@code DetailAST} ast element
+     * @return if {@code DetailAst} element is unique
+     */
+    private static boolean isUniqueAst(DetailAST ast) {
+        boolean isOnly = true;
+        final DetailAST identAst = ast.findFirstToken(TokenTypes.IDENT);
+
+        if (identAst == null) {
+            final DetailAST cur = getFirstSibling(ast);
```

Comment:
Why do we need this? We can use ast.getNextSibling(). Remove the method.
Do not create unnecessary local variable `identAst`.

```java
    private static boolean isUniqueAst(DetailAST ast) {
        return ast.findFirstToken(TokenTypes.IDENT) != null
            || !hasAtLeastOneSiblingWithSameTokenType(ast);
    }
```

After refactoring, the name of the method will be inappropriate. Rename the method.